New driver and other changes

FrostyMike opened the pull request:
## Description
Added build with pre-installed OMNI option
Switched to CMake based OMNI build
Added out of source build
Made tests run on install directory
Added explicit java installation option

Removed bash-based driver
Added java-based driver
Moved CMake-based tests to JUnit
Updated developer documentation
Updated sonar to new version
Added options for better support of GNU Make-based builds

clementval commented on documentation/developer/code/MyFirstTransformation.java:
```diff
@@ -1,6 +1,7 @@
 /*
  * This file is released under terms of BSD license
  * See LICENSE file for more information
+ * @author: not specified
```
Same comment about author. Put `clementval` or remove the line.

FrostyMike replied:
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

clementval commented on documentation/developer/code/basic_fortran.xml:
```diff
@@ -1,3 +1,8 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Same comment about author

FrostyMike replied:
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

clementval commented on documentation/developer/code/claw-default.xml:
```diff
@@ -1,3 +1,8 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Same comment about author

FrostyMike replied:
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

clementval commented on documentation/developer/code/claw-internal-set.xml:
```diff
@@ -1,3 +1,8 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Same comment about author

FrostyMike replied:
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

clementval commented on documentation/developer/code/clawp.g4:
```diff
@@ -1,3 +1,8 @@
+/*
+ * This file is released under terms of BSD license
+ * See LICENSE file for more information
+ * @author: not specified
```
Same comment about author

FrostyMike replied:
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

clementval commented on documentation/developer/code/extension_config.xml:
```diff
@@ -1,3 +1,8 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Same comment about author

FrostyMike replied:
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

clementval commented on documentation/developer/code/external-set.xml:
```diff
@@ -1,3 +1,8 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Same comment about author

FrostyMike replied:
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

clementval commented on documentation/developer/code/external_config.xml:
```diff
@@ -1,3 +1,8 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Same comment about author

FrostyMike replied:
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

clementval reviewed (changes requested):
I had a look at about 25% of the PR so far. Couple of general comments: 

* Lots of file have been moved in the directory structure but they appear as deleted/added in git. It must be corrected so it is a `git mv` and not a deletion/addition. Otherwise the history on the file is not preserve. 

* There are lots of `author not specified` added. In 99% of the file you can add `clementval` as the original author. Otherwise it doesn't make sense to add the tag with no author. 

I'll continue the review over the week-end.

clementval commented on scripts/run-test.sh:
```diff
@@ -0,0 +1,15 @@
+else
+    cat ${REPORT_FILE}
+    exit 1
+fi
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/cx2t.ant.builder.in:
```diff
@@ -0,0 +1,14 @@
+    <stringAttribute key="org.eclipse.ui.externaltools.ATTR_TOOL_ARGUMENTS"
+        value="-f @ANT_SOLUTION@ -Dclaw.properties.dir=@CX2T_PROPERTIES_DIR@ build-cx2t"/>
+    <booleanAttribute key="org.eclipse.ui.externaltools.ATTR_TRIGGERS_CONFIGURED" value="true"/>
+</launchConfiguration>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/cx2t.classpath.in:
```diff
@@ -0,0 +1,37 @@
+	</classpathentry>
+	<classpathentry kind="lib" path="@OMNI_JAR_TOOLS@"/>
+	<classpathentry kind="output" path="bin"/>
+</classpath>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/cx2t.project.in:
```diff
@@ -0,0 +1,39 @@
+			<location>@CX2T_GEN_SRC_DIR@</location>
+		</link>
+	</linkedResources>
+</projectDescription>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/clawfc.classpath.in:
```diff
@@ -0,0 +1,49 @@
+	<classpathentry kind="lib" path="@INT_CLAW_HOME@/bin/@CLAWFC_JAR@"/>
+
+	<classpathentry kind="output" path="bin"/>
+</classpath>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/clawfc.project.in:
```diff
@@ -0,0 +1,49 @@
+			<location>@CLAWFC_GEN_SRC_DIR@</location>
+		</link>
+	</linkedResources>
+</projectDescription>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/clawfc.ant.builder.in:
```diff
@@ -0,0 +1,14 @@
+    <stringAttribute key="org.eclipse.ui.externaltools.ATTR_TOOL_ARGUMENTS"
+        value="-f @ANT_SOLUTION@ -Dclaw.properties.dir=@CX2T_PROPERTIES_DIR@ main"/>
+    <booleanAttribute key="org.eclipse.ui.externaltools.ATTR_TRIGGERS_CONFIGURED" value="true"/>
+</launchConfiguration>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/clawfc-unit-tests.project.in:
```diff
@@ -0,0 +1,64 @@
+			<location>@CLAWFC_GEN_SRC_DIR@</location>
+		</link>
+	</linkedResources>
+</projectDescription>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/clawfc-unit-tests.ant.builder.in:
```diff
@@ -0,0 +1,14 @@
+    <stringAttribute key="org.eclipse.ui.externaltools.ATTR_TOOL_ARGUMENTS"
+        value="-f @ANT_SOLUTION@ -Dclaw.properties.dir=@CX2T_PROPERTIES_DIR@ build-driver-unit-tests"/>
+    <booleanAttribute key="org.eclipse.ui.externaltools.ATTR_TRIGGERS_CONFIGURED" value="true"/>
+</launchConfiguration>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/clawfc-tests.project.in:
```diff
@@ -0,0 +1,64 @@
+			<location>@CLAWFC_GEN_SRC_DIR@</location>
+		</link>
+	</linkedResources>
+</projectDescription>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/clawfc-tests.classpath.in:
```diff
@@ -0,0 +1,63 @@
+	<classpathentry kind="lib" path="@INT_CLAW_HOME@/bin/@CLAWFC_JAR@"/>
+
+	<classpathentry kind="output" path="bin"/>
+</classpath>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/claw-tests.ant.builder.in:
```diff
@@ -0,0 +1,14 @@
+    <stringAttribute key="org.eclipse.ui.externaltools.ATTR_TOOL_ARGUMENTS"
+        value="-f @ANT_SOLUTION@ -Dclaw.properties.dir=@CX2T_PROPERTIES_DIR@ build-claw-tests"/>
+    <booleanAttribute key="org.eclipse.ui.externaltools.ATTR_TRIGGERS_CONFIGURED" value="true"/>
+</launchConfiguration>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/claw-tests.classpath.in:
```diff
@@ -0,0 +1,58 @@
+	<classpathentry kind="lib" path="@INT_CLAW_HOME@/bin/@CLAWFC_JAR@"/>
+
+	<classpathentry kind="output" path="bin"/>
+</classpath>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/claw-tests.project.in:
```diff
@@ -0,0 +1,64 @@
+			<location>@CLAWFC_GEN_SRC_DIR@</location>
+		</link>
+	</linkedResources>
+</projectDescription>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on resources/eclipse/clawfc-tests.ant.builder.in:
```diff
@@ -0,0 +1,14 @@
+    <stringAttribute key="org.eclipse.ui.externaltools.ATTR_TOOL_ARGUMENTS"
+        value="-f @ANT_SOLUTION@ -Dclaw.properties.dir=@CX2T_PROPERTIES_DIR@ build-driver-tests"/>
+    <booleanAttribute key="org.eclipse.ui.externaltools.ATTR_TRIGGERS_CONFIGURED" value="true"/>
+</launchConfiguration>
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on driver/src/clawfc/BuildInfo.java:
```diff
@@ -0,0 +1,110 @@
+        return Collections.unmodifiableMap(res);
+    }
+
+};
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on driver/src/clawfc/BuildInfo.java:
```diff
@@ -0,0 +1,110 @@
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+public class BuildInfo
```
Can you add some comment to describe what this class is doing.

FrostyMike replied:
Added brief javadoc

clementval commented on driver/src/clawfc/BuildOrder.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.Map;
+import java.util.Set;
+
+public interface BuildOrder
```
Javadoc?

FrostyMike replied:
Added brief javadoc

clementval commented on driver/src/clawfc/CLAWAddIgnoreScanner.g4:
```diff
@@ -0,0 +1,58 @@
+fragment NOT_EOL_CHR :  ~[\r\n];
+
+fragment CLAW_IGNORE_DIR_START : SEP? '!' SEP? '$' CLAW SEP;
+fragment CLAW : C L A W;
```
Feels like overkill.

FrostyMike replied:
No comment

clementval commented on driver/src/clawfc/CLAWAddIgnoreScanner.g4:
```diff
@@ -0,0 +1,58 @@
+/**
+ * ANTLR 4 Grammar file for recognizing CLAW Ignore directives.
+ */
+grammar CLAWAddIgnoreScanner;
```
Why this is not part of the standard CLAW grammar.

FrostyMike replied:
Ignore directive is processed at the preprocessing stage, well before actual CLAW processing.

clementval commented on driver/src/clawfc/CLAWIgnoreLineRecognizer.g4:
```diff
@@ -0,0 +1,51 @@
+CLAW_IGNORE_DIR_END : SEP? ('\r')? '\n' ->skip;
+
+fragment NOT_EOL_CHR:  ~[\r\n];
+fragment CLAW : C L A W;
```
Duplicated and overkill. Same for the rest of the file.

FrostyMike replied:
Makes maintenance easier

clementval commented on driver/src/clawfc/CLAWRemoveIgnoreScanner.g4:
```diff
@@ -0,0 +1,58 @@
+fragment NOT_EOL_CHR:  ~[\r\n];
+
+fragment CLAW_IGNORE_DIR_START: SEP? '!' SEP? '$' CLAW SEP;
+fragment CLAW : C L A W;
```
Same comment as previous `g4` file.

FrostyMike replied:
Makes maintenance easier

clementval commented on driver/src/clawfc/CLAWVerbatimLineRecognizer.g4:
```diff
@@ -0,0 +1,53 @@
+EOL : ('\r')? '\n';
+
+fragment NOT_EOL_CHR:  ~[\r\n];
+fragment CLAW : C L A W;
```
Same comment as for other `g4` file.

FrostyMike replied:
Makes maintenance easier

clementval commented on driver/src/clawfc/CLAWVerbatimScanner.g4:
```diff
@@ -0,0 +1,53 @@
+fragment EOL : ('\r')? '\n';
+fragment NOT_EOL_CHR:  ~[\r\n];
+
+fragment CLAW : C L A W;
```
Same comment for `g4`

FrostyMike replied:
Makes maintenance easier

clementval commented on driver/src/clawfc/CommentTrailingBackslashRecognizer.g4:
```diff
@@ -0,0 +1,23 @@
+/**
+ * ANTLR 4 Grammar file for filtering out trailing backslashes at the end of Fortran comment.
+ */
+grammar CommentTrailingBackslashRecognizer;
```
Why do we need a grammar file for a simple stuff like that?

FrostyMike replied:
Makes maintenance easier. This way maintainer only knowledge of basic java and ANTLR would be required for the task.

clementval commented on driver/src/clawfc/Configuration.java:
```diff
@@ -0,0 +1,136 @@
+
+    Properties cfgProperties;
+    static final String OMNI_HOME_ENV = System.getenv("OMNI_HOME");
+};
```
NIT: missing new line.

FrostyMike replied:
Fixed

clementval commented on driver/src/clawfc/Driver.java:
```diff
@@ -0,0 +1,1545 @@
+                }
+                return;
+            }
+            if (opts.stopAfterXmodGeneration())
```
Remove empty if

FrostyMike replied:
Left on purpose, for debugging

clementval commented on driver/src/clawfc/Driver.java:
```diff
@@ -0,0 +1,1545 @@
+            }
+            if (opts.stopAfterXmodGeneration())
+            {
+            } else if (opts.stopAfterXastGeneration())
```
Remove  empty else if

FrostyMike replied:
It's like this on purpose, helps during debugging. As long as there's a line with opening bracket, I can put a breakpoint there.

clementval replied:
Conditional breakpoint should work fine for this purpose.

clementval commented on driver/src/clawfc/Driver.java:
```diff
@@ -0,0 +1,1545 @@
+                ClawX2T.verifyTargetOption(opts.targetPlatform());
+                ClawX2T.verifyConfigFile(opts.configFile());
+                ClawX2T.verifyModelConfigFile(opts.modelConfigFile());
+                if (opts.stopAfterTranslation())
```
Remove empty if

FrostyMike replied:
It's like this on purpose, helps during debugging. As long as there's a line with opening bracket, I can put a breakpoint there.

clementval commented on driver/src/clawfc/Driver.java:
```diff
@@ -0,0 +1,1545 @@
+        String vStr = sprintf("%s %s \"%s\" %s ", cfg().name(), cfg().version(), cfg().commit(), cfg().omniVersion());
+        print(vStr);
+    }
+};
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval commented on driver/src/clawfc/Driver.java:
```diff
@@ -0,0 +1,1545 @@
+                if (opts.stopAfterTranslation())
+                {
+
+                } else if (opts.stopAfterDecompilation())
```
Remove empty else if

FrostyMike replied:
It's like this on purpose. As long as there's a line there with opening bracket, I can put a breakpoint there.

clementval commented on driver/src/clawfc/FortranFileProgramUnitInfoData.java:
```diff
@@ -0,0 +1,227 @@
+        this.ts = ts;
+    }
+
+}
```
NIT: Missing new line

FrostyMike replied:
Fixed

clementval reviewed (changes requested):
About > 50% reviewed. Mostly simple comments.

clementval commented on driver/tests/res/translation/with_src_files/reference/mod11.xast:
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Is `.xast` extension introduced with the new driver?

FrostyMike replied:
".xast" is Fortran fully transformed to XCodeML. Introduced to distinguish from ".xmod". However, it is internal to CLAW.

clementval commented on driver/src/clawfc/utils/UniquePathHashGenerator.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Map;
+import java.util.Set;
+
+public class UniquePathHashGenerator extends SimplePathHashGenerator
```
Javadoc?

FrostyMike replied:
Added brief javadoc.

clementval commented on driver/src/clawfc/utils/UniquePathHashGenerator.java:
```diff
@@ -0,0 +1,50 @@
+        return hash;
+    }
+
+}
```
NIT: Missing new line.

FrostyMike replied:
Fixed

clementval commented on driver/src/clawfc/XmodData.java:
```diff
@@ -0,0 +1,95 @@
+        return outFilePath;
+    }
+
+}
```
Missing new line.

FrostyMike replied:
Fixed

clementval commented on driver/src/clawfc/depscan/FilteredContentSequence.java:
```diff
@@ -0,0 +1,179 @@
+import java.util.List;
+import java.util.Map;
+
+public class FilteredContentSequence
```
Javadoc?

FrostyMike replied:
Added brief javadoc.

clementval commented on driver/src/clawfc/depscan/FortranDepStatementsRecognizer.java:
```diff
@@ -0,0 +1,379 @@
+import clawfc.depscan.parser.FortranDepStatementsRecognizerLexer;
+import clawfc.depscan.parser.FortranDepStatementsRecognizerParser;
+
+public class FortranDepStatementsRecognizer implements FortranProgramUnitStatementsRecognizer
```
Javadoc

FrostyMike replied:
Added brief javadoc.

clementval commented on driver/src/clawfc/depscan/FortranException.java:
```diff
@@ -0,0 +1,59 @@
+ */
+package clawfc.depscan;
+
+public class FortranException extends Exception
```
Javadoc

FrostyMike replied:
Added brief javadoc.

clementval commented on driver/src/clawfc/depscan/FortranIncludesResolver.g4:
```diff
@@ -0,0 +1,56 @@
+fragment SQ : '\'';
+fragment INCLUDE : I N C L U D E;
+
+fragment A : [aA];
```
Most of fragment are not used. Lexer rule can be defined in an easier way.

FrostyMike replied:
File is tiny anyway

clementval replied:
If you want to keep unused fragments for letters at least centralize them in a single and you can used these lexer rules in all your files. 

```
lexer grammar MyLexer;
```

```
options { tokenVocab= MyLexer; }
```

clementval reviewed (changes requested):
I went through all the files. Thanks for the PR. I think overall it will improve the quality of the driver and the debugging possibility. 

In the last couple of files I reviewed I saw couple of things that should be improved. 

* In the `g4` ANTLR file. Fragment are not used correctly to define lexer rule. Lexer rules can be defined in an easier way. Lots of fragment are not used or duplicated. A single file of lexer rules can be done. 
* Javadoc is missing in most of newly added file. We do not expect a javadoc description on each getter and setter but at least on the class declaration with a brief description of the purpose of the class.
* Couple of file have not been moved correctly with git so they appear as added/deleted.

FrostyMike commented:
> * Lots of file have been moved in the directory structure but they appear as deleted/added in git. It must be corrected so it is a `git mv` and not a deletion/addition. Otherwise the history on the file is not preserve.

This branch was created mostly by copying over the files with filesystem, so some git-specific info was lost. "Fixing" this will require more time and effort that I am prepared to spend on this. I see 2 possibilities here:
1. This PR stays as it is, as far as moved files are concerned. From lots of github "File renamed without changes." messages, I gather that in 90% of cases git duly detected that file was moved.
2. This PR will be closed and we revert back to to the [original PR](https://github.com/claw-project/claw-compiler/pull/600). It has identical code, but different git history, where all the files were indeed renamed with "git mv". A side effect of this, however, will be that [coding style PR](https://github.com/claw-project/claw-compiler/pull/603) will have to be closed, and changes to the coding style will have to be done after merging https://github.com/claw-project/claw-compiler/pull/600 , instead of before.

I am fine with either option. You can decide.

> * There are lots of `author not specified` added. In 99% of the file you can add `clementval` as the original author. Otherwise it doesn't make sense to add the tag with no author.

Again, unfortunately this involves spending a lot of time on something that does not improve the end user experience or make maintenance easier (which are the goals of my work). The purpose of "author not specified" is to silence the warning from the author check script. The message means that original file did not contain an "author" field as required by the guidelines.

If you wish, you can go over the files in question, and put in the correct author.

FrostyMike commented:
> I went through all the files. Thanks for the PR. I think overall it will improve the quality of the driver and the debugging possibility.

Thanks for your review. I'll go over individual comments next week.

> * In the `g4` ANTLR file. Fragment are not used correctly to define lexer rule. Lexer rules can be defined in an easier way. Lots of fragment are not used or duplicated. A single file of lexer rules can be done.

I am well aware that the way I used ANTLR is not the most efficient one. However, in this case the priority was mostly the ease of maintenance. While most likely it was possible to do everything in one or several big grammars, it would make it significantly more complicated. I decided early on to have many smaller grammars instead. Of course this involved a lot of code duplication, but the advantage is that they are self-contained and easier to read. 

> * Javadoc is missing in most of newly added file. We do not expect a javadoc description on each getter and setter but at least on the class declaration with a brief description of the purpose of the class.

Due to resource constraints it is unfortunately not possible to document everything. In most cases I opt for self-descriptive names instead. "FortranCLAWDetector" and "This class detects use of CLAW in Fortran code" amount to approximately the same, only the first one is shorter.

clementval commented:
> 1. This PR stays as it is, as far as moved files are concerned. From lots of github "File renamed without changes." messages, I gather that in 90% of cases git duly detected that file was moved.
> 2. This PR will be closed and we revert back to to the [original PR](https://github.com/claw-project/claw-compiler/pull/600). It has identical code, but different git history, where all the files were indeed renamed with "git mv". A side effect of this, however, will be that [coding style PR](https://github.com/claw-project/claw-compiler/pull/603) will have to be closed, and changes to the coding style will have to be done after merging #600 , instead of before.
> 
> I am fine with either option. You can decide.

Ok then let's keep it like that. 

> Again, unfortunately this involves spending a lot of time on something that does not improve the end user experience or make maintenance easier (which are the goals of my work). The purpose of "author not specified" is to silence the warning from the author check script. The message means that original file did not contain an "author" field as required by the guidelines.
> 
> If you wish, you can go over the files in question, and put in the correct author.

I'll do it. It takes literally 5 seconds with an IDE or a `sed` command.

clementval commented:
> I am well aware that the way I used ANTLR is not the most efficient one. However, in this case the priority was mostly the ease of maintenance. While most likely it was possible to do everything in one or several big grammars, it would make it significantly more complicated. I decided early on to have many smaller grammars instead. Of course this involved a lot of code duplication, but the advantage is that they are self-contained and easier to read.
> 

Fine with the multiple grammar files but I was talking about the way lexer rules are defined in each grammar file. This should and can be simplified a lot. 

> Due to resource constraints it is unfortunately not possible to document everything. In most cases I opt for self-descriptive names instead. "FortranCLAWDetector" and "This class detects use of CLAW in Fortran code" amount to approximately the same, only the first one is shorter.

At least a brief description for the class would be nice. We do not need each methods to be documented of course.

FrostyMike commented on driver/src/clawfc/depscan/FortranLineBreaksFilter.g4:
```diff
@@ -0,0 +1,29 @@
+EOL : ('\r')? '\n';
+
+fragment SEP : WS+;
+fragment WS : [ \t];
```
Fixed

FrostyMike commented on driver/src/clawfc/depscan/PreprocessorLineMarkerRecognizer.g4:
```diff
@@ -0,0 +1,27 @@
+
+fragment SEP : WS+;
+fragment WS : [ \t];
+fragment EOL : '\r'? '\n';
```
Fixed (newline)

FrostyMike commented on driver/src/clawfc/depscan/PreprocessorOutputSplitter.g4:
```diff
@@ -0,0 +1,32 @@
+
+fragment SEP : WS+;
+fragment WS : [ \t];
+fragment EOL : '\r'? '\n';
```
Fixed (newline)

FrostyMike commented on compiler/cray.cmake:
```diff
@@ -16,3 +16,4 @@ set(OPENACC_FLAGS "-h acc,noomp")     # flags to compile with OpenACC support
 set(OPENMP_FLAGS "-h noacc,omp")      # flags to compile with OpenMP support
 set(OMNI_TARGET "")                   # Compilation of OMNI compiler
 set(COMPILER_MACRO "-D_CRAYFTN")      # predefined macro by compiler
+set(FC_VENDOR "cray")
```
Fixed

FrostyMike commented on compiler/intel.cmake:
```diff
@@ -15,3 +15,4 @@ set(TEST_BASE_FLAGS "")       # Base flags for test case compilation
 set(OPENACC_FLAGS "")         # flags to compile with OpenACC support
 set(OPENMP_FLAGS "-openmp")   # flags to compile with OpenMP support
 set(COMPILER_MACRO "-D__ICC")    # predefined macro by compiler
+set(FC_VENDOR "intel")
```
Fixed

clementval commented:
Is this PR still going to grow? 

@lxavier Do we have confirmation that this works with the application used at MCH and ECMWF? 

On ORNL side, we are currently testing it. We will report the result shortly.

FrostyMike commented:
> Is this PR still going to grow?

I've run into (more) problems during testing on Piz Daint. Since attempts to fix them in CLAW driver were time-consuming and ineffective, I decided to focus my attention on OMNI compiler instead. After OMNI is updated, I will do corresponding update to this PR. This is the only planned change, unless fresh bugs are found.

For the purposes of testing, I recommend using the last revision that is not failing. I will likely rollback 
"Reimplemented Subprocess.call with Unix shell" 8fb9c42.

FrostyMike commented on compiler/pgi.cmake:
```diff
@@ -17,3 +17,4 @@ set(OPENACC_FLAGS "-acc")               # flags to compile with OpenACC support
 set(OPENMP_FLAGS "-mp")                 # flags to compile with OpenMP support
 set(OMNI_CONF_OPTION ${OMNI_CONF_OPTION} "CPP=pgcc -E")
 set(COMPILER_MACRO "-D_PGI")            # predefined macro by compiler
+set(FC_VENDOR "portland")
```
For legacy reasons it has to remain portland

FrostyMike commented on cx2t/src/claw/config/claw-external-set.xml:
```diff
@@ -0,0 +1,13 @@
+<!--
```
Too much work required. See https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808712088

FrostyMike replied:
Too much work required. See https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808712088

FrostyMike commented on cx2t/src/claw/config/claw-high-level-set.xml:
```diff
@@ -0,0 +1,24 @@
+<!--
```
Too much work required. See https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808712088

FrostyMike commented on cx2t/src/claw/config/claw-internal-set.xml:
```diff
@@ -0,0 +1,13 @@
+<!--
```
Too much work required. See https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808712088

FrostyMike commented on cx2t/src/claw/config/claw-low-level-set.xml:
```diff
@@ -0,0 +1,56 @@
+<!--
```
Too much work required. See https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808712088

FrostyMike commented on cx2t/src/claw/config/claw-default.xml:
```diff
@@ -0,0 +1,142 @@
+<!--
```
Too much work required. See https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808712088

FrostyMike commented on test/loops/nested_hoist/reference.f90:
```diff
@@ -0,0 +1,28 @@
+PROGRAM LOOP_FUSION
```
Too much work required. See https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808712088

FrostyMike commented on cx2t/src/claw/config/claw_transformation_set.xsd:
```diff
@@ -0,0 +1,42 @@
+<!--
```
Too much work required. See https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808712088

FrostyMike commented on driver/src/clawfc/depscan/FortranCLAWScanner.g4:
```diff
@@ -0,0 +1,56 @@
+fragment NOT_EOL_CHR:  ~[\r\n];
+
+fragment ACC : A C C;
+fragment CLAW : C L A W;
```
Easier to maintain.

FrostyMike commented on driver/src/clawfc/depscan/FortranCLAWScanner.java:
```diff
@@ -0,0 +1,121 @@
+import clawfc.depscan.parser.FortranCLAWScannerLexer;
+import clawfc.depscan.parser.FortranCLAWScannerParser;
+
+public class FortranCLAWScanner
```
File removed

FrostyMike commented on driver/src/clawfc/depscan/FortranDepScanner.g4:
```diff
@@ -0,0 +1,144 @@
+fragment INTEGER : I N T E G E R;
+fragment REAL : R E A L;
+
+fragment BIND : B I N D;
```
Works fine.

FrostyMike commented on driver/src/clawfc/depscan/FortranDepStatementsRecognizer.g4:
```diff
@@ -0,0 +1,105 @@
+PROGRAM  |
+USE;
+
+BLOCKDATA : B L O C K D A T A;
```
Works fine

FrostyMike commented on driver/src/clawfc/depscan/FortranIncludeStatementRecognizer.g4:
```diff
@@ -0,0 +1,54 @@
+fragment DQ : '"';
+fragment SQ : '\'';
+
+fragment A : [aA];
```
File is tiny anyway

FrostyMike commented on driver/src/clawfc/depscan/FortranProcedureStatementsRecognizer.g4:
```diff
@@ -0,0 +1,119 @@
+
+DOUBLE_PRECISION : DOUBLE SEP PRECISION;
+
+fragment CHARACTER : C H A R A C T E R;
```
Works fine and easier to maintain this way.

FrostyMike commented on driver/src/clawfc/FortranFrontEnd.java:
```diff
@@ -0,0 +1,218 @@
+        return driverIntDir.resolve("xmods");
+    }
+
+    public FortranFrontEnd(Configuration cfg, Options opts, Path driverTempDir) throws Exception
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/IgnoreDirectiveRecognizer.java:
```diff
@@ -0,0 +1,123 @@
+import clawfc.parsers.CLAWIgnoreLineRecognizerParser;
+import clawfc.utils.ParserErrorListener;
+
+public class IgnoreDirectiveRecognizer
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/Options.java:
```diff
@@ -0,0 +1,729 @@
+        return parseCmdlineArguments(args, Utils.STARTUP_DIR);
+    }
+
+    public static Options parseCmdlineArguments(String[] args, Path workingDir) throws Exception
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/PreprocessedFortranSourceData.java:
```diff
@@ -0,0 +1,121 @@
+import clawfc.depscan.FortranFileProgramUnitInfo;
+import clawfc.utils.AsciiArrayIOStream;
+
+public class PreprocessedFortranSourceData
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/Preprocessor.java:
```diff
@@ -0,0 +1,334 @@
+import clawfc.utils.Subprocess;
+import clawfc.utils.SubprocessFailed;
+
+public class Preprocessor
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/ProgramUnitData.java:
```diff
@@ -0,0 +1,236 @@
+import clawfc.depscan.serial.FortranProgramUnitType;
+import clawfc.utils.AsciiArrayIOStream;
+
+public class ProgramUnitData implements ProgramUnitInfo
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/RemoveIgnoreDirectiveFilter.java:
```diff
@@ -0,0 +1,148 @@
+import clawfc.utils.AsciiArrayIOStream;
+import clawfc.utils.ParserErrorListener;
+
+public class RemoveIgnoreDirectiveFilter
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/RemoveVerbatimDirectiveFilter.java:
```diff
@@ -0,0 +1,149 @@
+import clawfc.utils.AsciiArrayIOStream;
+import clawfc.utils.ParserErrorListener;
+
+public class RemoveVerbatimDirectiveFilter
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/Utils.java:
```diff
@@ -0,0 +1,402 @@
+import clawfc.utils.FileInfoImpl;
+import clawfc.utils.Subprocess;
+
+public class Utils
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/VerbatimDirectiveRecognizer.java:
```diff
@@ -0,0 +1,123 @@
+import clawfc.parsers.CLAWVerbatimLineRecognizerParser;
+import clawfc.utils.ParserErrorListener;
+
+public class VerbatimDirectiveRecognizer
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/depscan/FortranCLAWDetector.java:
```diff
@@ -0,0 +1,111 @@
+import clawfc.depscan.parser.FortranCLAWScannerLexer;
+import clawfc.depscan.parser.FortranCLAWScannerParser;
+
+public class FortranCLAWDetector
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/depscan/FortranCommentsFilter.java:
```diff
@@ -0,0 +1,144 @@
+import clawfc.depscan.parser.FortranCommentsFilterLexer;
+import clawfc.depscan.parser.FortranCommentsFilterParser;
+
+public class FortranCommentsFilter
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/depscan/FortranDepParser.java:
```diff
@@ -0,0 +1,599 @@
+import clawfc.depscan.parser.FortranDepScannerParser;
+import clawfc.depscan.serial.FortranProgramUnitType;
+
+public class FortranDepParser
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/depscan/FortranDepScanner.java:
```diff
@@ -0,0 +1,259 @@
+import clawfc.utils.AsciiArrayIOStream;
+import clawfc.utils.ByteArrayIOStream;
+
+public class FortranDepScanner
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/TrailingBackslashCommentsFilter.java:
```diff
@@ -0,0 +1,178 @@
+import clawfc.depscan.parser.FortranCommentsFilterParser;
+import clawfc.utils.AsciiArrayIOStream;
+
+public class TrailingBackslashCommentsFilter
```
Added brief javadoc.

FrostyMike commented on driver/src/clawfc/ProgramUnitInfo.java:
```diff
@@ -0,0 +1,63 @@
+import clawfc.utils.AsciiArrayIOStream;
+import clawfc.depscan.serial.FortranProgramUnitType;
+
+public interface ProgramUnitInfo
```
Added brief javadoc.

FrostyMike commented on cx2t/ivy.xml:
```diff
@@ -1,6 +1,7 @@
 <!--
  This file is released under terms of BSD license
  See LICENSE file for more information
+ @author: not specified
```
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

FrostyMike commented on cx2t/src/build.xml:
```diff
@@ -1,32 +1,28 @@
 <!--
  This file is released under terms of BSD license
  See LICENSE file for more information
+ @author: not specified
```
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

FrostyMike commented on cx2t/src/claw/config/claw-default.xml:
```diff
@@ -0,0 +1,142 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

FrostyMike commented on cx2t/src/claw/config/claw-external-set.xml:
```diff
@@ -0,0 +1,13 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

FrostyMike commented on cx2t/src/claw/config/claw-high-level-set.xml:
```diff
@@ -0,0 +1,24 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

FrostyMike commented on cx2t/src/claw/config/claw-internal-set.xml:
```diff
@@ -0,0 +1,13 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

FrostyMike commented on cx2t/src/claw/config/claw-low-level-set.xml:
```diff
@@ -0,0 +1,56 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

FrostyMike commented on cx2t/src/claw/config/claw_config.xsd:
```diff
@@ -0,0 +1,69 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

FrostyMike commented on cx2t/src/claw/config/claw_transformation_set.xsd:
```diff
@@ -0,0 +1,42 @@
+<!--
+ This file is released under terms of BSD license
+ See LICENSE file for more information
+ @author: not specified
```
Required now, to silence automatic check. As [discussed ](https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808755861) you can put in the author manually yourself.

FrostyMike commented on cx2t/src/claw/config/claw_config.xsd:
```diff
@@ -0,0 +1,69 @@
+<!--
```
Too much work required. See https://github.com/claw-project/claw-compiler/pull/604#issuecomment-808712088

FrostyMike commented:
Update:
- fixed problems with OMNI compiler;
- fixed an issue with "use" statetement recognition;
- added CLI options --max-num-mp-jobs and --ffront-debug-dir;
- added brief java doc to clawfc classes;

lxavier reviewed (approved):
You can merge when tests are green

lxavier commented:
Changes were tested with both cosmo and icon
